### DISCUSSION TITLE
sqlmodel `StrEnum` vs. `Enum` processing

### DISCUSSION THREAD

**Discussion by danmurphy1217** [sqlalchemy/sqlalchemy#12123] [Usage Questions]
We have the following Str Enums:

```
class ChequeUserStatus(enum.StrEnum):
	ACTIVE = "Active"
        ...

class ChequeAccountStatus(enum.StrEnum):
	ACTIVE = "Active"
        ...
```
we typed the columns in our respective tables as such:

```
...cheque_accounts...
vendor_status: ChequeAccountStatus | None = base.EnumStrField(
    ChequeAccountStatus, nullable=True
)

...cheque_users...
vendor_status: ChequeUserStatus | None = base.EnumStrField(
    ChequeUserStatus, nullable=True
)
```

`base.EnumStrField` is defined as this:

```

def EnumStrField(enum: Any, **kwargs: Any) -> Any:
    if "sa_type" in kwargs:
        raise ValueError("sa_type is not allowed for EnumStrField")

    return Field(
        sa_type=EnumAsString(enum_class=enum),  # type: ignore
        **kwargs,
    )

 
# Define a custom type decorator to serialize the Enum as a string
class EnumAsString(TypeDecorator):  # type: ignore
    """
    Customer type to serialize the Enum as a string,
    and deserialize the string back to the Enum.
    """

    impl = AutoString
    cache_ok = True

    def __init__(self, enum_class: type[PyEnum] | None = None, *args, **kwargs) -> None:  # type: ignore
        # NOTE: for auto-gen migrations we support None
        self.enum_class = enum_class
        super().__init__(*args, **kwargs)

    def process_bind_param(self, value: PyEnum | None, dialect: Any) -> str | None:
        if value is None:
            return None
        elif isinstance(value, str):
            # ensure valid enum
            self.enum_class(value)  # type: ignore
            return value
        return value.value  # type: ignore

    def process_result_value(self, value: str, dialect: Any) -> PyEnum | None:
        if value is not None:
            return self.enum_class(value)  # type: ignore
        return None
```

when querying for data from our tables, we get the following error:

```
File "/opt/pysetup/.venv/lib/python3.12/site-packages/sqlalchemy/sql/sqltypes.py", line 1611, in _object_value_for_elem
raise LookupError(
LookupError: 'Active' is not among the defined enum values. Enum name: chequeaccountstatus. Possible values: ACTIVE, INACTIVE...
```

The full stacktrace is:

```
/sqlalchemy/engine/result.py:1810 - one()
/sqlalchemy/engine/result.py:749 - _only_one_row
/sqlalchemy/engine/result.py:1673 - _fetchone_impl
/sqlalchemy/engine/result.py:2259 - _fetchone_impl
/sqlalchemy/orm/loading.py:219 - chunks
/sqlalchemy/engine/result.py:541 - _raw_all_rows
/sqlalchemy/sql/sqltypes.py:1729 - process
/sqlalchemy/sql/sqltypes.py:1611 - _object_value_for_elem
```

Changing the typing of the enums to `Enum` solved the problem. It _seems_ like there is an issue on how `StrEnums` vs. Traditional `Enums` are processed/handled.

Is this a known issue? Is there something specific to our implementation that might be incorrect?

I followed thru the stacktrace and it seems like the issue could be with how the "valid_lookup" is performed in sqlalchemy:

```
    def _setup_for_values(self, values, objects, kw):
        self.enums = list(values)

        self._valid_lookup = dict(zip(reversed(objects), reversed(values)))

        self._object_lookup = dict(zip(values, objects))

        self._valid_lookup.update(
            [
                (value, self._valid_lookup[self._object_lookup[value]])
                for value in values
            ]
        )
```

**Comment by CaselIT** [comment]
Hi,

It's really hard to follow what you are doing. What is Field, what is AutoString?
There is no usage of the sqlalchemy Enum in your code, so I don't know how that comes in play.

**Comment by danmurphy1217** [comment]
Hi @CaselIT, i've included the relevant imports below:

```
from enum import Enum as PyEnum
from typing import Any

from sqlalchemy import TypeDecorator
from sqlmodel import AutoString, Field

def EnumStrField(enum: Any, **kwargs: Any) -> Any:
    if "sa_type" in kwargs:
        raise ValueError("sa_type is not allowed for EnumStrField")

    return Field(
        sa_type=EnumAsString(enum_class=enum),  # type: ignore
        **kwargs,
    )

 
# Define a custom type decorator to serialize the Enum as a string
class EnumAsString(TypeDecorator):  # type: ignore
    """
    Customer type to serialize the Enum as a string,
    and deserialize the string back to the Enum.
    """

    impl = AutoString
    cache_ok = True

    def __init__(self, enum_class: type[PyEnum] | None = None, *args, **kwargs) -> None:  # type: ignore
        # NOTE: for auto-gen migrations we support None
        self.enum_class = enum_class
        super().__init__(*args, **kwargs)

    def process_bind_param(self, value: PyEnum | None, dialect: Any) -> str | None:
        if value is None:
            return None
        elif isinstance(value, str):
            # ensure valid enum
            self.enum_class(value)  # type: ignore
            return value
        return value.value  # type: ignore

    def process_result_value(self, value: str, dialect: Any) -> PyEnum | None:
        if value is not None:
            return self.enum_class(value)  # type: ignore
        return None
```

both `Field` and `AutoString` are imported from **sqlmodel**:

```
from sqlmodel import AutoString, Field
```

> There is no usage of the sqlalchemy Enum in your code, so I don't know how that comes in play.

My hypothesis was formed primarily by following the stacktrace and digging into some of the details of how enums are handled by sqlalchemy. Hopefully you can shine a bit more light on this for us 😄

**Reply by CaselIT** [reply]
Thanks for providing more details.
This suggest that whatever sqlmodel is doing it not correct for StrEnum. I would suggest asking there.
This is an example using only sqlalchemy:
```py
import enum
import sqlalchemy as sa

engine = sa.create_engine("sqlite://", echo=True)

class E(enum.StrEnum):
    foo = "bar"

t = sa.Table("t", sa.MetaData(), sa.Column("c", sa.Enum(E)))

t.metadata.create_all(engine)
with engine.connect() as c:
    c.execute(t.insert(), {"c": E.foo})
    print(c.execute(t.select()).one())
```

**Reply by csm-kb** [reply]
In case folks get stuck here in the future looking to use enum values instead of keys, after some extra digging in other discussions:
- You want to use the `values_callable` kwarg in sqlalchemy's `Enum` constructor like so:
```py
enum MyEnum(StrEnum):
    TheKey = "the.value"

class Model(...):
    column: Mapped[MyEnum] = mapped_column(sqlalchemy.types.Enum(MyEnum, values_callable=lambda x: [str(e.value) for e in x]), ...) # x is MyEnum

# inserts will store "the.value" when referenced as TheKey in Python land, and convert nicely to TheKey when queried
```

Context ref this [doc link](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Enum.params.values_callable) as of date:
> A callable which will be passed the PEP-435 compliant enumerated type, which should then return a list of string values to be persisted. This allows for alternate usages such as using the string value of an enum to be persisted to the database instead of its name. The callable must return the values to be persisted in the same order as iterating through the `Enum`’s `__member__` attribute. For example `lambda x: [i.value for i in x]`.

Re [this discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/8934), that doc page is likely the best support we're going to get for enum values of arbitrary types in the library!
